# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: netwerk/protocol/http/HttpBackgroundChannelParent.cpp
# Commit: 3b268ac50692
# Full Hash: 3b268ac50692bae6b1bcd0c41e61e267a6735e68
# Author: Junior Hsu <juhsu@mozilla.com>
# Date: 2020-06-24 09:31:07
# Regressor Bug: 1633935
# File Overlap Count: 1
# Description:
#   Bug 1633935 - P9 send OnAfterLastPart via pBg IPC to avoid race between OS*R, r=mayhemer,necko-reviewers
#   
#   Depends on D77750
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D77751
# ==============================================================================

diff -r ac9c5c1e5162 -r 3b268ac50692 netwerk/protocol/http/HttpBackgroundChannelParent.cpp
--- a/netwerk/protocol/http/HttpBackgroundChannelParent.cpp	Tue Jun 23 16:58:29 2020 +0000
+++ b/netwerk/protocol/http/HttpBackgroundChannelParent.cpp	Tue Jun 23 16:58:33 2020 +0000
@@ -239,6 +239,30 @@
                            aResponseTrailers, aConsoleReports);
 }
 
+bool HttpBackgroundChannelParent::OnAfterLastPart(const nsresult aStatus) {
+  LOG(("HttpBackgroundChannelParent::OnAfterLastPart [this=%p]\n", this));
+  AssertIsInMainProcess();
+
+  if (NS_WARN_IF(!mIPCOpened)) {
+    return false;
+  }
+
+  if (!IsOnBackgroundThread()) {
+    MutexAutoLock lock(mBgThreadMutex);
+    nsresult rv = mBackgroundThread->Dispatch(
+        NewRunnableMethod<const nsresult>(
+            "net::HttpBackgroundChannelParent::OnAfterLastPart", this,
+            &HttpBackgroundChannelParent::OnAfterLastPart, aStatus),
+        NS_DISPATCH_NORMAL);
+
+    MOZ_DIAGNOSTIC_ASSERT(NS_SUCCEEDED(rv));
+
+    return NS_SUCCEEDED(rv);
+  }
+
+  return SendOnAfterLastPart(aStatus);
+}
+
 bool HttpBackgroundChannelParent::OnProgress(const int64_t aProgress,
                                              const int64_t aProgressMax) {
   LOG(("HttpBackgroundChannelParent::OnProgress [this=%p]\n", this));